src/symbolic_dimensions.jl

```diff
@@ -95,14 +99,118 @@ end
 expand_units(q::QuantityArray) = expand_units.(q)
 
 
-Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(data(d)))
-Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions) = data(l) == data(r)
-Base.iszero(d::SymbolicDimensions) = iszero(data(d))
-Base.:*(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) + data(r))
-Base.:/(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) - data(r))
-Base.inv(d::SymbolicDimensions) = SymbolicDimensions(-data(d))
-Base.:^(l::SymbolicDimensions{R}, r::Integer) where {R} = SymbolicDimensions(data(l) * r)
-Base.:^(l::SymbolicDimensions{R}, r::Number) where {R} = SymbolicDimensions(data(l) * tryrationalize(R, r))
+Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(getfield(d, :nzdims)), copy(getfield(d, :nzvals)))
+function Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions)
+    nzdimsl = getfield(l, :nzdims)
+    nzvalsl = getfield(l, :nzvals)
+    nzdimsr = getfield(r, :nzdims)
+    nzvalsr = getfield(r, :nzvals)
+    nl = length(nzdimsl)
+    nr = length(nzdimsr)
+    il = ir = 1
+    while il <= nl && ir <= nr
+        diml = nzdimsl[il]
+        dimr = nzdimsr[ir]
+        if diml == dimr
+            if nzvalsl[il] != nzvalsr[ir]
+                return false
+            end
+            il += 1
+            ir += 1
+        elseif diml < dimr
```

Comment:
Not sure I understand. What do these branches for `diml < dimr` and `diml > dimr` do?

Reply:
Oh wait, I think I get it.

Wouldn't it be simpler (both conceptually and perhaps computationally) to first pop any 0s from each vector, and then we only need to check `nzdimsl == nzdimsr` and `nzvalsl == nzvalsr`?

Reply:
Popping elements will cause unnecessary allocations (if I understood your suggestion correctly). This should be the most efficient algorithm since it only loops over the dimensions once without allocating anything and with exiting early. The same algorithm is also used in SparseArrays: https://github.com/JuliaSparse/SparseArrays.jl/blob/d884072ee95d522505c4b27fa928dc3c251f7d2b/src/sparsevector.jl#L937-L966

Reply:
I see, thanks!